R/utils-get_code_dependency.R

```diff
@@ -0,0 +1,491 @@
+#' @return A `list` (of length of input `calls_pd`) where each element represents one call. Each element consists of a
+#' character vector containing names of objects that were influenced by this call, and names of objects influencing this
+#' call. Influencers appear after `":"` string, e.g. `c("a", ":", "b", "c")` means a call influenced object named `a`,
+#' and the object was influenced by objects named `b` and `c`. If an object was marked with `@linksto` side effects tag
+#' then it appears as an influenced object at the beginning of a vector.
+#'
+#' @keywords internal
+#' @noRd
+code_graph <- function(calls_pd) {
+
+  cooccurence <- extract_occurrence(calls_pd)
+
+  side_effects <- extract_side_effects(calls_pd)
+
+  graph <- mapply(c, side_effects, cooccurence, SIMPLIFY = FALSE)
+
+  remove_graph_duplicates(graph)
+
+}
+
+#' Extract Object Occurrence
+#'
+#' @description Extract objects occurrence within calls passed by `calls_pd`. It also detects which objects are
+#' influenced by others within a call, and which are influencers.
+#'
+#' @param calls_pd A `list` of `data.frame`s, which is a result of `utils::getParseData()` grouped into separate calls.
+#' A result of `extract_calls()` function.
+#' @return A `list` (of length of input `calls_pd`) where each element represents one call. Each element consists of a
+#' character vector containing names of objects that were influenced by this call, and names of objects influencing this
+#' call. Influencers appear after `":"` string, e.g. `c("a", ":", "b", "c")` means a call influenced object named `a`,
+#' and the object was influenced by objects named `b` and `c`.
+#' @keywords internal
+#' @noRd
+extract_occurrence <- function(calls_pd) {
+
+  used_in_function <- function(call, rows) {
+
+    # If an object is a function parameter,
+    # then in calls_pd there is a `SYMBOL_FORMALS` entry for that object.
+    objects <- sapply(rows, function(row) call$text[row])
+
+    vapply(objects, function(object) {
+      if (any(call[call$token == "SYMBOL_FORMALS", "text"] == object) && any(call$token == "FUNCTION")) {
+        object_sf_ids <- call[call$text == object & call$token == "SYMBOL", "id"]
+        function_start_id <- call[call$token == "FUNCTION", "id"]
+        all(object_sf_ids > function_start_id)
+      } else {
+        FALSE
+      }
+    },
+      logical(1)
+    )
+  }
+
+  extract_function_names <- function(calls_pd) {
+    # Returns function names which are created within code and called.
+    pd <- do.call(rbind, calls_pd)
+    symbols <- pd[pd$token %in% c("SYMBOL", "SYMBOL_FUNCTION_CALL"), c('token', 'text')]
+    symbols_table <- table(unique(symbols)$text)
+    names(symbols_table[symbols_table == 2])
+  }
+
+  functions <- extract_function_names(calls_pd)
+
+  lapply(
+    calls_pd,
+    function(x) {
+      sym_cond <- which(x$token == "SYMBOL" | (x$token == "SYMBOL_FUNCTION_CALL" & x$text %in% functions))
+      sym_cond <- sym_cond[!used_in_function(x, sym_cond)]
```

Comment:
Please take `get_children` from `extract_calls` and you'll have a nice simplification

```suggestion
  is_in_function <- function(x) {
    # If an object is a function parameter,
    # then in calls_pd there is a `SYMBOL_FORMALS` entry for that object.
    function_id <- x[x$token == "FUNCTION", "parent"]
    if (length(function_id)) {
      x$id %in% get_children(x, function_id)$id
    } else {
      rep(FALSE, nrow(x))
    }
  }

  lapply(
    calls_pd,
    function(call_pd) {
      # what occurs in a function body is not tracked
      x <- call_pd[!is_in_function(call_pd), ]
      sym_cond <- which(x$token %in% c("SYMBOL", "SYMBOL_FUNCTION_CALL"))
```

Reply:
https://github.com/insightsengineering/teal.data/pull/201/commits/3480f9847b487553402a576fc18e62994d028514